### DISCUSSION TITLE
How to add device via udev in linux?

### DISCUSSION THREAD

**Discussion by AmeyaPatwardhan** [pymeasure/pymeasure#505] [Q&A] [answered]
I wanted to use pymeasure to connect to an oscilloscope over usb. I found the idVendor and idProduct using lsusb. However, when I try to create a device using udev, with the SUBSYSTEM=="usb", SYMLINK+="oscilloscope", it creates a symlink to  /dev/bus/..... and not to ttyUSBn. I tried using SUBSYSTEM=='tty", but it fails to create any device in this case. Help would be much appreciated

**Comment by LongnoseRob** [comment, marked as answer]
Hello,
It might be helpful to tell us brand and type of the oscilloscope.

Assuming you are on Linux (as you are talking about udev), have you checked if your kernel supports USB-TMC as this is what many modern oscilloscopes support nowadays?

There is a possibilty to use pyvisa and pyvisa-py with pyusb to access USB instruments: 
[pyvisa-py docu](https://pyvisa.readthedocs.io/projects/pyvisa-py/en/latest/installation.html#usb-resources-usb-instr-raw)
[pyvisa docu](https://pyvisa.readthedocs.io/en/latest/introduction/communication.html)

If your system does not support USB-TMC out of the box, you might try this [experimental driver](https://github.com/dpenkler/linux-usbtmc)

**Reply by bilderbuchi** [reply]
Also find examples for VISA resource strings for USB at https://pyvisa.readthedocs.io/en/latest/introduction/names.html

**Reply by AmeyaPatwardhan** [reply]
Output from lsusb, 
Bus 001 Device 010: ID 0699:0362 Tektronix, Inc. Tektronix TDS1001B

**Reply by LongnoseRob** [reply]
Then I would recommend to have a look at [this](https://itectec.com/ubuntu/ubuntu-how-to-connect-tektronix-tds10001b-to-ubuntu/)

**Reply by AmeyaPatwardhan** [reply]
Thanks for the help,  I managed to get the oscilloscope working after a bit of work. I can use the python shell to execute all commands and queries on the instrument. However, when I try to use the Procedure class alongside the MainWindow, I get the following error





Worker caught an error on <MeasureFluxProcedure(status=Running,parameters_are_set=True)> (ERROR)
Traceback (most recent call last):
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pyvisa_py/protocols/usbtmc.py", line 256, in write
    return self.usb_send_ep.write(data)
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/usb/core.py", line 408, in write
    return self.device.write(self, data, timeout)
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/usb/core.py", line 989, in write
    return fn(
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/usb/backend/libusb1.py", line 837, in bulk_write
    return self.__write(self.lib.libusb_bulk_transfer,
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/usb/backend/libusb1.py", line 938, in __write
    _check(retval)
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/usb/backend/libusb1.py", line 602, in _check
    raise USBTimeoutError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBTimeoutError: [Errno 110] Operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pymeasure/experiment/workers.py", line 165, in run
    self.procedure.startup()
  File "beamstability.py", line 69, in startup
    log.debug(self.osc.ask(':HOR?').strip())
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pymeasure/instruments/instrument.py", line 92, in ask
    return self.adapter.ask(command)
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pymeasure/adapters/visa.py", line 112, in ask
    return self.connection.query(command)
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pyvisa/resources/messagebased.py", line 638, in query
    self.write(message)
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pyvisa/resources/messagebased.py", line 197, in write
    count = self.write_raw(message.encode(enco))
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pyvisa/resources/messagebased.py", line 157, in write_raw
    return self.visalib.write(self.session, message)[0]
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pyvisa_py/highlevel.py", line 543, in write
    written, status_code = self.sessions[session].write(data)
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pyvisa_py/usb.py", line 179, in write
    count = self.interface.write(data)
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pyvisa_py/protocols/usbtmc.py", line 436, in write
    bytes_sent += raw_write(data)
  File "/home/ameya/anaconda3/envs/pymeasure/lib/python3.8/site-packages/pyvisa_py/protocols/usbtmc.py", line 258, in write
    raise ValueError(str(e))
ValueError: [Errno 110] Operation timed out
10/30/2021 09:21:51 PM : Finished Acquisition (INFO)
10/30/2021 09:21:51 PM : Monitor caught stop command (INFO)



Here is the relevant code that I am using


class MeasureFluxProcedure(Procedure):

    iterations = IntegerParameter('Loop Iterations', default=10)
    delay = FloatParameter('Delay Time', units='s', default=1)

    DATA_COLUMNS = ['Time', 'MeanV',]



    def __init__(self):
        super(MeasureFluxProcedure, self).__init__()
        log.info("Initializing instrument")
        # inital default setup

    def startup(self):
        
        self.resman = pyvisa.ResourceManager('@py')
        self.TDS1001B_address = "USB0::1689::866::C050712::0::INSTR"
        
        if self.TDS1001B_address not in self.resman.list_resources():    
            logging.error("Could not connect to TDS 1001 B")
            raise Exception("Could not connect to TDS 1001 B")

        self.adpt = VISAAdapter(self.TDS1001B_address, visa_library='@py')
        self.osc = TDS2000(self.adpt)

        time.sleep(1)
        
        
        self.osc.write("*CLS; *WAI")
        self.osc.ask("*IDN?; *WAI")
        self.osc.write('*RST; *WAI; :AUTOS EXEC; *WAI')


        #horizontal scaling
        self.osc.write(':HOR:DEL:POS 0.0E0; MAI:POS:0.0E0; SCA %.2E; :HOR:RECO 2500; *WAI' %float(10e-3) ) # 10 ms, recordlength not relevant for TDS 1001 B
        log.debug(self.osc.ask(':HOR?').strip())

        #triggering setup
        self.osc.write(':TRIG:MAI:MOD AUTO; *WAI')
        log.debug(self.osc.ask(':TRIG:MAI?'.strip()))

        #vertical scaling setup for channel 1
        self.osc.write(':CH1:BAN OFF; COUP DC; POS 0e0; SCA 20e-3; *WAI') # 20 mV lowest permitted by TDS 1001B
        log.debug(self.osc.ask(':CH1?').strip())

        #setting up measurement parameters
        # self.osc.write(':MEASU:MEAS1:SOU CH1; :MEASU:MEAS2:SOU CH1; :MEASU:MEAS3:SOU CH1; :MEASU:MEAS4:SOU CH1; :MEASU:MEAS5:SOU CH1;')
        # self.osc.write(':MEASU:MEAS1:TYP MEAN; :MEASU:MEAS2:TYP MINI; :MEASU:MEAS3:TYP MAXI; :MEASU:MEAS4:TYP NON; :MEASU:MEAS5:TYP NON;')
        # log.debug(self.osc.ask(':MEASU:MEAS1?; :MEASU:MEAS2?; :MEASU:MEAS3?; :MEASU:MEAS4?; :MEASU:MEAS5?;').strip())
        self.osc.write(":MEASU:IMM:SOU1 CH1; TYP MEAN; *WAI")
        log.debug(self.osc.ask(":MEASU:IMM?").strip())
        log.debug(self.osc.ask(":MEASU:IMM:VAL?").strip())
        


    def execute(self):
        start_time = time.time()
        last_measurement = start_time
        log.info("Starting Acquisition")
        for i in range(int(self.iterations)):
            data = {
                'Time': time.time()-start_time,
                'MeanV': float(self.osc.ask(':MEASU:IMM:VAL?; *WAI').strip()),
                # 'MinV': float(self.osc.ask(':MEASU:MEAS2:VAL?').strip()),
                # 'MaxV': float(self.osc.ask(':MEASU:MEAS3:VAL?').strip()),
            }
            self.emit('results', data)
            self.emit('progress', 100*i/self.iterations)
            time.sleep(float(self.delay))
            last_measurement = last_measurement + self.delay

            if self.should_stop():
                log.warning("Catch stop command in procedure")
                break
        


    def shutdown(self):
        # self.osc.write("*WAI")
        # self.osc.shutdown()
        # self.adpt.connection.close()
        log.info("Finished Acquisition")
        

class MainWindow(ManagedWindow):

    def __init__(self):
        super(MainWindow, self).__init__(
            procedure_class=MeasureFluxProcedure,
            inputs=['iterations', 'delay'],
            displays=['iterations', 'delay'],
            x_axis='Time',
            y_axis='MeanV',
        )
        self.setWindowTitle('GUI Example')

    def queue(self):

        filename = tempfile.mktemp()
        procedure = self.make_procedure()
        results = Results(procedure,filename)
        experiment = self.new_experiment(results)
        self.manager.queue(experiment)


if __name__ == "__main__":
    app = QtGui.QApplication(sys.argv)
    window = MainWindow()
    window.show()
    sys.exit(app.exec_())

**Comment by LongnoseRob** [comment]
I am not 100% but this 
```
self.osc.write(':HOR:DEL:POS 0.0E0; MAI:POS:0.0E0; SCA %.2E; :HOR:RECO 2500; *WAI' %float(10e-3) ) # 10 ms, recordlength not relevant for TDS 1001 B
```
Could it be that the 2nd or 3rd commands in this line are not acepted? 
What happens if you change it to `:HOR:MAI:POS 0.0E0; :HOR:SCA %.2E;`?

it might be also helpful to use `self.osc.check_errors()` to see if the instrument reports any errors

**Reply by AmeyaPatwardhan** [reply]
Thanks. Works fine now!

**Reply by msmttchr** [reply]
@AmeyaPatwardhan could you please point out what was the problem and the solution, so we can mark this as answered?

**Reply by AmeyaPatwardhan** [reply]
Done
